group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/topics/CopartitionedTopicsEnforcerTest.java

```diff
@@ -0,0 +1,261 @@
+            TopicConfigurationException.class,
+            () -> enforcer.enforce(
+                Set.of(repartitionTopic1, repartitionTopic2),
+                Set.of(repartitionTopic1, repartitionTopic2),
+                Set.of()
+            )
+        );
+
+        final TreeMap<String, Integer> sorted = new TreeMap<>(
+            Utils.mkMap(Utils.mkEntry(repartitionTopic1, 10),
+                Utils.mkEntry(repartitionTopic2, 5))
+        );
+        assertEquals(Status.INCORRECTLY_PARTITIONED_TOPICS, ex.status());
+        assertEquals(String.format(
+            "Following topics do not have the same number of partitions: " +
+                "[%s]", sorted), ex.getMessage());
+    }
+
+    @Test
+    public void shouldNotThrowAnExceptionWhenTopicInfosWithEnforcedNumOfPartitionsAreValid() {
+        final String repartitionTopic1 = "repartitioned-1";
+        final String repartitionTopic2 = "repartitioned-2";
+        final Map<String, Integer> topicPartitionCounts = Map.of(
+            repartitionTopic1, 10,
+            repartitionTopic2, 10
+        );
+        final CopartitionedTopicsEnforcer enforcer =
+            new CopartitionedTopicsEnforcer(LOG_CONTEXT, topicPartitionProvider(topicPartitionCounts));
+
+        final Map<String, Integer> enforced = enforcer.enforce(
+            Set.of(repartitionTopic1, repartitionTopic2),
+            Set.of(),
+            Set.of(repartitionTopic1, repartitionTopic2)
+        );
+
+        assertEquals(Map.of(
+            repartitionTopic1, 10,
+            repartitionTopic2, 10
+        ), enforced);
+    }
+
+    @Test
+    public void shouldThrowAnExceptionWhenNumberOfPartitionsOfNonRepartitionTopicAndRepartitionTopicWithEnforcedNumOfPartitionsDoNotMatch() {
+        final String repartitionTopic1 = "repartitioned-1";
+        final String firstSourceTopic = "first";
+        final Map<String, Integer> topicPartitionCounts = Map.of(
+            repartitionTopic1, 10,
+            firstSourceTopic, 2
+        );
+        final CopartitionedTopicsEnforcer enforcer =
+            new CopartitionedTopicsEnforcer(LOG_CONTEXT, topicPartitionProvider(topicPartitionCounts));
+
+        final TopicConfigurationException ex = assertThrows(
+            TopicConfigurationException.class,
+            () -> enforcer.enforce(
+                Set.of(repartitionTopic1, firstSourceTopic),
+                Set.of(repartitionTopic1),
+                Set.of())
+        );
+
+        assertEquals(Status.INCORRECTLY_PARTITIONED_TOPICS, ex.status());
+        assertEquals(String.format("Number of partitions [%s] " +
+                "of repartition topic [%s] " +
+                "doesn't match number of partitions [%s] of the source topic.",
+            10, repartitionTopic1, 2), ex.getMessage());
+    }
+
+    @Test
+    public void shouldNotThrowAnExceptionWhenNumberOfPartitionsOfNonRepartitionTopicAndRepartitionTopicWithEnforcedNumOfPartitionsMatch() {
```

Comment:
Also here could you please try to formulate this positively?

Reply:
Done